core/ui/src/main/java/org/phoebus/ui/application/PhoebusApplication.java

```diff
@@ -1263,22 +1293,347 @@ private void closeMainStage() {
+                    var applicationNameToDockItemsWithInput = windowNrToApplicationNameToDockItemsWithInput.get(windowName);
+
+                    if (applicationNameToDockItemsWithInput.size() > 0) {    // Only print unsaved changes for a window if it actually containts any unsaved changes.
+                        if (windowNrToApplicationNameToDockItemsWithInput.size() >= 2) {    // Only print the window names if two or more windows are in the process of being closed.
+                            Text windowTitle = new Text(windowName);
+                            windowTitle.setStyle("-fx-font-size: 16; -fx-font-weight: bold");
+                            gridPane.add(windowTitle, 0, currentRow);
+                            currentRow++;
+                        }
+
+                        for (var applicationName : applicationNameToDockItemsWithInput.keySet()) {
+                            for (var dockItemWithInput : applicationNameToDockItemsWithInput.get(applicationName)) {
+                                CheckBox checkBox = new CheckBox();
+                                checkBox.selectedProperty().addListener((observableValue, old_value, new_value) -> enableAndDisableButtons.run());
+
+                                Text applicationName_text = new Text(applicationName + ":");
+                                applicationName_text.setStyle("-fx-font-weight: bold");
+                                Text instanceName_text = new Text(dockItemWithInput.getLabel());
+
+                                HBox hBox = new HBox(checkBox, applicationName_text, instanceName_text);
+                                hBox.setSpacing(4);
+                                gridPane.add(hBox, 0, currentRow);
+
+                                Consumer<Boolean> setCheckboxStatus = bool -> checkBox.setSelected(bool);
+                                setCheckBoxStatusActions.add(setCheckboxStatus);
+
+                                Supplier<Boolean> getCheckBoxStatus = () -> checkBox.isSelected();
+                                getCheckBoxStatusActions.add(getCheckBoxStatus);
+
+                                hBox.addEventHandler(MouseEvent.MOUSE_CLICKED, mouseEvent -> checkBox.setSelected(!checkBox.isSelected())); // Enable toggling checkbox by clicking on its label.
+
+                                Supplier<SaveStatus> saveIfCheckboxEnabled = () -> {
+                                    if (checkBox.isSelected()) {
+
+                                        Text saving = new Text("[" + Messages.UnsavedChanges_saving + "]");
+                                        saving.setFill(Color.ORANGE);
+                                        saving.setStyle("-fx-font-weight: bold;");
+                                        hBox.getChildren().set(0, saving);
+                                        boolean saveSuccessful = dockItemWithInput.save(monitor, prompt.getDialogPane().getScene().getWindow());
```

Comment:
Is this call to dockItemWithInput.save() blocking? (The return type boolean suggests it is, but I do not know what the parameter monitor does.)